doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+# PREVIEW: New `libelektra-core` API 2023
+
+## Core API
+
+### Note on names
+
+From the [use cases](/doc/usecases/core/README.md) we know that there is a collection structure which contains elements with a name and value (and metadata).
+
+The current API uses these names for those concepts:
+
+- Collection: `KeySet`
+- Element: `Key`
+- Name of element: —
+- Value of element: —
+
+The API proposed below uses (to keep as much as possible from above):
+
+- Collection: `ElektraKeyset`
+- Element: `ElektraKey`
+- Name of element: `ElektraKeyname`
+- Value of element: `ElektraKeyvalue`
+
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
+- Name of element: `ElektraKey` or `ElektraName`
+- Value of element: `ElektraValue`
+
+### Note on public structs
+
+The public structs `ElektraKeyname` and `ElektraKeyvalue` defined below are not problematic or limiting in terms of forward compatibility.
+This is because, they essentially define what a "keyname" and a "key value" are for Elektra.
+That is why they are used both as part of the public API **and** in the definitions of the `*Cow` structs, which define how stuff is actually stored.
+Any change to these structs would mean changing that definition and therefore would be major breaking change, even if the structs where not public.
+
+### Public headers
+
+```c
+// indicates special rules for constness, see comments at use-site for details
+#define ELEKTRA_CONST const
+
+typedef uint8_t ElektraNamespace;
+
+typedef struct {
+    // constness follows containing struct
+    // if the containing struct is a `const ElektraKeyname *` then this field must not be modified
+    // if the containing struct is a `ElektraKeyname *` or a non-pointer `ElektraKeyname` it is okay to modify
+    ELEKTRA_CONST char * name;
+    size_t size;
+    ElektraNamespace ns;
+} ElektraKeyname;
+
+typedef struct {
+    const void * value;
+    size_t size;
+} ElektraKeyvalue;
+
+typedef struct ElektraKey ElektraKey;
+typedef struct ElektraKeyset ElektraKeyset;
+
+// 0 always indicates success, other values depend on function
+typedef int ElektraErrorCode;
+
+// Note: breaking change compared to current values
```

Comment:
Not really relevant for pre-1.0

```suggestion
```

Reply:
Yeah, I know I just added at for completeness's sake.